pkg/engine2/edge_targets.go

```diff
@@ -0,0 +1,297 @@
+The returned map is keyed by the source resource's string representation.
+The value for each source resource is a list of valid target resources.
+
+Targets are considered valid if there is a set of kb paths between the source and target
+that satisfies both source and target path satisfaction classifications.
+
+A partial set of valid targets can be generated using the filter criteria in the context's config.
+*/
+func (e *Engine) GetValidEdgeTargets(context *GetPossibleEdgesContext) (map[string][]string, error) {
+
+	inputGraph, err := unmarshallInputGraph(context.InputGraph)
+	if err != nil {
+		return nil, err
+	}
+	solutionCtx := NewSolutionContext(e.Kb)
+	err = solutionCtx.LoadGraph(inputGraph)
+	if err != nil {
+		return nil, err
+	}
+	topologyGraph, err := e.GetViewsDag(DataflowView, solutionCtx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var sources []construct.ResourceId
+	var targets []construct.ResourceId
+
+	qualifiedTypeMatcher := func(id construct.ResourceId) func(otherType construct.ResourceId) bool {
+		return func(otherType construct.ResourceId) bool {
+			return otherType.QualifiedTypeName() == id.QualifiedTypeName()
+		}
+	}
+
+	// filter resources based on the context
+	err = construct.WalkGraph(topologyGraph, func(id construct.ResourceId, resource *construct.Resource, nerr error) error {
+		tag := e.GetResourceVizTag(string(DataflowView), id)
+		if len(context.Tags) > 0 && !slices.Contains(context.Tags, tag) {
+			return nil
+		}
+		isSource := true
+		isTarget := true
+		if len(context.Resources.Sources) > 0 && !slices.Contains(context.Resources.Sources, id) {
+			isSource = false
+		}
+		if len(context.Resources.Targets) > 0 && !slices.Contains(context.Resources.Targets, id) {
+			isTarget = false
+		}
+
+		if len(context.ResourceTypes.Sources) > 0 && !slices.ContainsFunc(context.ResourceTypes.Sources, qualifiedTypeMatcher(id)) {
+			isSource = false
+		}
+
+		if len(context.ResourceTypes.Targets) > 0 && !slices.ContainsFunc(context.ResourceTypes.Targets, qualifiedTypeMatcher(id)) {
+			isTarget = false
+		}
+
+		if isSource {
+			sources = append(sources, id)
+		}
+		if isTarget {
+			targets = append(targets, id)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	results := make(chan *edgeValidity)
+
+	//var detectionGroup sync.WaitGroup
+
+	checkerPool := pond.New(5, 1000, pond.Strategy(pond.Lazy()))
+	knownTargetValidity := make(map[string]map[string]bool)
+	rwLock := &sync.RWMutex{}
+
+	// get all valid-edge combinations for resource types in the supplied graph
+	for _, s := range sources {
+		for _, t := range targets {
+			source := s
+			target := t
+
+			if source.Matches(target) {
+				continue
+			}
+
+			if source.Namespace == target.Name || target.Namespace == source.Name {
+				continue
+			}
+
+			path, err := graph.ShortestPath(topologyGraph, source, target)
+			if len(path) > 0 && err == nil {
+				continue
+			}
```

Comment:
sucks to have to run this just to determine if theres a path that already exists because theres definitely more efficient ways.